allennlp/commands/predict.py

```diff
@@ -94,62 +102,95 @@ def _get_predictor(args: argparse.Namespace) -> Predictor:
-            if output_file:
-                output_file.write(string_output)
-
-    batch_json_data = []
-    for line in input_file:
-        if not line.isspace():
-            # Collect batch size amount of data.
-            json_data = predictor.load_line(line)
-            batch_json_data.append(json_data)
-            if len(batch_json_data) == batch_size:
-                _run_predictor(batch_json_data)
-                batch_json_data = []
-
-    # We might not have a dataset perfectly divisible by the batch size,
-    # so tidy up the scraps.
-    if batch_json_data:
-        _run_predictor(batch_json_data)
+            print("prediction: ", prediction)
+        if self._output_file is not None:
+            self._output_file.write(prediction)
+
+    def _get_json_data(self) -> Iterator[JsonDict]:
+        for line in open(self._input_file):
+            if not line.isspace():
+                yield self._predictor.load_line(line)
+
+    def _get_instance_data(self) -> Iterator[Instance]:
+        if self._dataset_reader is None:
+            raise ConfigurationError("To generate instances directly, pass a DatasetReader.")
+        else:
+            yield from self._dataset_reader.read(self._input_file)
+
+    def run(self) -> None:
+        has_reader = self._dataset_reader is not None
+        if has_reader:
+            for batch in lazy_groups_of(self._get_instance_data(), self._batch_size):
+                for result in self._predict_instances(batch):
+                    self._maybe_print_to_console_and_file(result)
+        else:
+            for batch_json in lazy_groups_of(self._get_json_data(), self._batch_size):
+                for model_input, result in zip(batch_json, self._predict_json(batch_json)):
+                    self._maybe_print_to_console_and_file(result, json.dumps(model_input))
 
+        if self._output_file is not None:
+            self._output_file.close()
 
 def _predict(args: argparse.Namespace) -> None:
     predictor = _get_predictor(args)
-    output_file = None
 
     if args.silent and not args.output_file:
         print("--silent specified without --output-file.")
         print("Exiting early because no output will be created.")
         sys.exit(0)
 
-    # ExitStack allows us to conditionally context-manage `output_file`, which may or may not exist
-    with ExitStack() as stack:
-        input_file = stack.enter_context(args.input_file)  # type: ignore
-        if args.output_file:
-            output_file = stack.enter_context(args.output_file)  # type: ignore
-
-        _run(predictor,
-             input_file,
-             output_file,
-             args.batch_size,
-             not args.silent)
+    util = _Predict(predictor,
+                    args.input_file,
+                    args.output_file,
+                    args.batch_size,
+                    not args.silent,
+                    args.use_dataset_reader)
+    util.run()
```

Comment:
related to above, `util` isn't the most helpful name, maybe if it were `predictor_wrapper` or `predictor_runner` or something